satpy/composites/__init__.py

```diff
@@ -1714,24 +1726,51 @@ def _combine_metadata_with_mode_and_sensor(self,
 
     @staticmethod
     def _get_merged_image_data(foreground: xr.DataArray,
-                               background: xr.DataArray
+                               background: xr.DataArray,
+                               initial_bg_alpha: bool,
                                ) -> list[xr.DataArray]:
-        if "A" in foreground.attrs["mode"]:
-            # Use alpha channel as weight and blend the two composites
-            alpha = foreground.sel(bands="A")
-            data = []
-            # NOTE: there's no alpha band in the output image, it will
-            # be added by the data writer
-            for band in foreground.mode[:-1]:
-                fg_band = foreground.sel(bands=band)
-                bg_band = background.sel(bands=band)
-                chan = (fg_band * alpha + bg_band * (1 - alpha))
-                chan = xr.where(chan.isnull(), bg_band, chan)
-                data.append(chan)
-        else:
-            data_arr = xr.where(foreground.isnull(), background, foreground)
-            # Split to separate bands so the mode is correct
-            data = [data_arr.sel(bands=b) for b in data_arr["bands"]]
+        def _get_alpha(dataset: xr.DataArray):
+            # If the dataset contains an alpha channel, just use it
+            # If not, we still need one. So build it and fill it with 1
+            if "A" in dataset.attrs["mode"]:
+                alpha = dataset.sel(bands="A")
+            else:
+                first_band = dataset.isel(bands=0)
+                alpha = xr.full_like(first_band, 1)
+                alpha["bands"] = "A"
+
+            # There could be Nans in the alpha, especially through 'add_bands'
```

Comment:
In what cases does `add_bands` add an alpha with NaNs as described in this comment? Skimming the `add_bands` code I only see a `da.ones` call to create a fully opaque alpha channel.

Reply:
If this line got disabled I'll get a weird result.
![image](https://github.com/pytroll/satpy/assets/72339781/724a5f29-bd34-43c9-8b54-01ff8f1ad3fe)

Reply:
Ok. I would think other alpha-based composites (like some of the cloud ones you've talked about in this PR) are producing NaNs in the alpha. `add_bands` doesn't seem to be doing that.

Reply:
I'll change the comment there.